specs/payment-api.yaml

```diff
@@ -936,7 +936,17 @@ components:
           schema:
             type: array
             items:
-              $ref: '#/components/schemas/DebitNoteEvent'
+              anyOf:
+                - $ref: '#/components/schemas/DebitNoteReceivedEvent'
+                - $ref: '#/components/schemas/DebitNoteAcceptedEvent'
+                - $ref: '#/components/schemas/DebitNoteRejectedEvent'
+                - $ref: '#/components/schemas/DebitNoteFailedEvent'
+                - $ref: '#/components/schemas/DebitNoteSettledEvent'
+                - $ref: '#/components/schemas/DebitNoteCancelledEvent'
+                - $ref: '#/components/schemas/DebitNotePaymentStatusEvent'
+                - $ref: '#/components/schemas/DebitNotePaymentOkEvent'
+              discriminator:
+                propertyName: eventType
```

Comment:
I'm not knowledgeable enough to answer whether the OAS specification intends to remove the method of specifying payload variants.

However, I am certain that the change is implemented in line with the recommended way of documenting polymorphic response variants, as outlined in https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/.

This change enables our API client generator to resolve the return types for event listing endpoints. I believe it's perfectly fine :)

Reply:
> I'm not knowledgeable enough to answer whether the OAS specification intends to remove the method of specifying payload variants.

I believe there's been some miscommunication. I'm not worried about a future version of OAS dropping support for this feature, I'm asking whether the clients compliant with this schema will be able to parse new events if (when) we add them. For example, if we introduce `DebitNoteFooEvent`, will old clients be able to parse that?

I believe the answer is no, and that's the reason our schema is the way it is ATM. Please correct me if I'm wrong.

Reply:
I see what you mean now :)

>  I'm asking whether the clients compliant with this schema will be able to parse new events if (when) we add them. For example, if we introduce `DebitNoteFooEvent`, will old clients be able to accept that?

I think this always depends on the OAS-client-generator implementation and how strict they are. I can only answer for the one that we're using: `openapi-typescript-codegen`.

In our case, the generated client will receive the new event type from the API and pass it to the user of the client - there's no event type validation at runtime done by the generated client, so the _old clients will be able to accept that_ and the responsibility to handle "newly added stuff" correctly is on the user of the client. Essentially the generated client is a "statically typed pass-thru" without any additional intelligence.

Now, the _application using the client_ can be written in a bad way, which will make it prone to new event types (if it's not written in a forward-compatible way).

Our application (golem-js SDK) will:

* receive this new event (pass-thru) from the generated client,
* see a different event type than the expected one
* and ignore it (we are waiting for specific events, the rest is dropped).

Thus, we're safe (as JS SDK), and ya-ts-client users are safe as well in case you add new event types.

### On backward compatibility

In general, when you add more types of events like `DebitNoteFoo` that means that you're altering the response of the API which always bears the risk of being BC-breaking. _The discussion starts from the yagna (this is where the real BC break happens) endpoint and not the OAS specification._

So if you change the endpoint in yagna to produce new types of events there are few possibilities:

1. API consumers that are written in a strict way (non-forward compatible) might be hit when you produce a new event type from the API and their code will crash because of that event - parsing error, unsupported event type etc. The fault is on the consumer for being too strict here.

2. API consumers that are written in a forward-compatible way will accept the new event type returned from Yagna and will operate without disruption - the consumer will ignore the newly added event and continue.

3. API consumers are written in a "syntax-wise forward compatible way", you add a new event type but also introduce change in the semantics of the events. For example: previously you had events `A, B, C` that represented a complete multi-step process. If you add event `D` and then change the semantics of "the process" to be `A,B,C,D` and not `A,B,C` anymore that's a BC-breaking change on its own.

In `golem-js` we're using `openapi-typescript-codegen` and apply practices that position us in option **3**. We can accommodate any new event types unless they don't change the semantics which can break our implementation. 

But that's just for context. Not really relevant in context of this PR :)